Adds support for system account lookup

bkatyl opened the pull request:
Submission containing materials of a third party: @NoMoreFood

This contribution is from @NoMoreFood's uncommitted branch. https://github.com/NoMoreFood/openssh-portable/commit/64d4c840dc6ae986c6f5c1a150c0348843367bea.

This is part 1 of the a 2 part PR to Fix issue #1546

bkatyl committed:
Adds support for system account lookup

Submission containing materials of a third party: @NoMoreFood

This contribution is from @NoMoreFood's uncommitted branch. https://github.com/NoMoreFood/openssh-portable/commit/64d4c840dc6ae986c6f5c1a150c0348843367bea.

This is part 1 of the a 2 part PR to Fix issue #1546

bagajjal commented on contrib/win32/win32compat/pwd.c:
```diff
@@ -236,9 +236,11 @@ get_passwd(const wchar_t * user_utf16, PSID sid)
-	if ((_wcsicmp(domain_name, computer_name) == 0) && (_wcsicmp(computer_name, user_name) != 0))
-		wcscpy_s(user_resolved, ARRAYSIZE(user_resolved), user_name);
+	/* if standard local user name or system account, just use name without decoration */
+	const SID_IDENTIFIER_AUTHORITY nt_authority = SECURITY_NT_AUTHORITY;
```
1)If statement is missing body. 
wcscpy_s(user_resolved, ARRAYSIZE(user_resolved), user_name);

looks like the build failed because of this.

2) please use () as there are multiple &&, || conditions here. It improves readability.

bkatyl replied:
Sorry about missing the body, it was a copy/paste error between branches.

Great point about using additional () to increase the readability.

bagajjal commented on contrib/win32/win32compat/pwd.c:
```diff
@@ -236,9 +236,11 @@ get_passwd(const wchar_t * user_utf16, PSID sid)
 		goto cleanup;
 	}
 
```
Please add steps to verify the new functionality.

bkatyl replied:
To verify the functionality set the following in the ssh_config
AuthorizedKeysCommandUser=system

I'm looking at how to extend the regress\pesterTests\AuthorizedKeysCommand.Tests.ps1 test to allow the AuthorizedKeysCommandUser to be specified instead of always using $ssouser. So then there would be 2 runs of the tests, one with the existing $ssouser and the other as system.

bkatyl replied:
I'm not sure why the following test failed in the appveyor build test.
```Run transfer.sh [9 of 71] 03/18/2021 16:39:38
transfer.sh FAILED 03/18/2021 16:42:05
```

I ran the test suite against the combined changes and also this specific change by using the same commands run to test only this change in this [test run](https://ci.appveyor.com/project/PowerShell/openssh-portable/builds/38293371) and they succeeded.

```
Run transfer.sh [9 of 71] 03/19/2021 15:14:05
transfer.sh PASSED 03/19/2021 15:14:18
...
{
    "StartTime":  "3/19/2021 3:12:37 PM",
    "EndTime":  "3/19/2021 4:06:50 PM",
    "TotalExecutionTime":  "00:54:12",
    "TotalBashTests":  71,
    "TotalBashTestsPassed":  64,
    "TotalBashTestsFailed":  0,
    "TotalBashTestsSkipped":  7,
    "FailedBashTests":  "",
    "SkippedBashTests":  "agent.sh, key-options.sh, cert-hostkey.sh, cert-userkey.sh, forward-control.sh, integrity.sh, krl.sh",
    "BashTestSummaryFile":  "C:\\projects\\openssh-portable\\regress\\bash_tests_summary.txt",
    "BashTestLogFile":  "C:\\projects\\openssh-portable\\regress\\bash_tests_output.log"
}
```

catthehacker replied:
From log:
```
Run transfer.sh [9 of 71] 03/18/2021 16:39:38
generating key type ssh-ed25519
generating key type ssh-rsa
generating key type ssh-dss
generating key type ecdsa-sha2-nistp256
generating key type ecdsa-sha2-nistp384
generating key type ecdsa-sha2-nistp521
dd-size 10
ssh cat /cygdrive/c/projects/openssh-portable/regress/temp_test/data failed
cmp: /cygdrive/c/projects/openssh-portable/regress/temp_test/copy: No such file or directory
corrupted copy
dd-size 100
dd-size 1k
dd-size 32k
dd-size 64k
dd-size 128k
dd-size 256k
failed transfer data
transfer.sh FAILED 03/18/2021 16:42:05
```

bagajjal replied:
@bkatyl, I did minor changes and resolved the merge conflicts in pwd.c. My intention is  to restart the build as it appears transfer.sh randomly failed .

Hope you don't mind.

bkatyl replied:
That's great, thank you. It looks like all the tests now pass. :)

bkatyl committed:
Adding if statement body and () for readability

- Adding () to improve readability.
- Fixing copy/paste error between branches, by adding if statement body.

bagajjal committed:
Update pwd.c

bagajjal committed:
resolve merge conflicts after fetching latest changes

illfelder commented:
It looks like tests are finally passing! Great job @bkatyl! 

Thank you for the thorough reviews @bagajjal - please let us know if any other changes are needed to merge this or #479
